addons/udes_stock/models/stock_picking.py

```diff
@@ -296,57 +332,74 @@ def _backorder_move_lines(self):
-        unfulfilled_moves = moves.filtered(
-            lambda mv: sum(mv.move_line_ids.mapped("qty_done")) < mv.product_uom_qty
-        )
+        # Create a record for the moves to be moved into the backorder
+        new_moves = Move.browse()
 
-        # Raise an error if:
-        # * there are moves or move lines that are not fully fulfilled so we can backorder
-        # something.
-        # * if move lines need splitting
-        if not unfulfilled_moves:
-            raise ValidationError(
-                _("There are no move lines within picking %s to backorder") % self.name
-            )
-        elif mls and any(ml.qty_done != 0 and ml.qty_done < ml.product_uom_qty for ml in mls):
-            raise ValidationError(
-                _("You cannot create a backorder for %s with a move line qty less than expected!")
-                % self.name
-            )
+        # Determine the mls to keep if not provided
+        if not mls_to_keep:
+            mls_to_keep = incomplete_moves.move_line_ids.get_lines_done()
 
-        # Iterate over the moves not fulfilled, and split out recording the newly created moves
-        new_moves = Move.browse()
-        for move in unfulfilled_moves:
-            new_moves |= move.split_out_incomplete_move()
+        # If any moves are untouched by mls_to_keep add them directly to new_moves
+        touched_moves = mls_to_keep.mapped("move_id")
+        untouched_moves = incomplete_moves - touched_moves
+        new_moves |= untouched_moves
+
+        # Check a backorder is allowed, we don't want to move work completed out, if
+        # the original picking is not done.
+        self._assert_backorder_allowed(mls_to_keep, untouched_moves)
+
+        for move in touched_moves.get_uncovered_moves(mls_to_keep):
+            rel_mls = mls_to_keep & move.move_line_ids
+            new_moves |= move.split_out_move(mls_to_keep=rel_mls)
+
+        # Create backorder picking for remaining moves and return
+        if new_moves:
+            return self._create_backorder_picking(new_moves)
 
-        # Create picking for completed move
+        # If no bakcorder return an empty record set
+        return Picking.browse()
+
+    def _create_backorder_picking(self, moves):
+        """Helper to create a backorder picking from the given moves"""
         bk_picking = self.copy(
             {
                 "name": "/",
-                "move_lines": [(6, 0, new_moves.ids)],
-                "move_line_ids": [(6, 0, new_moves.move_line_ids.ids)],
+                "move_lines": [(6, 0, moves.ids)],
+                "move_line_ids": [(6, 0, moves.move_line_ids.ids)],
                 "backorder_id": self.id,
             }
         )
-
         return bk_picking
 
-    def _requires_backorder(self, mls):
-        """Checks if a backorder is required by checking if all move lines
-        within a picking are present in mls
+    def _requires_backorder(self, mls=None):
```

Comment:
Nit: it is not very clear which conditions means that the backorder is required or not. Can we make it a little clearer?